### PR TITLE
Remove System.Collections.Immutable dependency

### LINKED ISSUE #1040: New MVC App + CloudTrace.Initialize() yields exception.
I created a new MVC web app.  Added the dependency Google.Cloud.Diagnostics.AspNet.  Then added a call to CloudTrace.Initialize(), and I get this exception when I try to run the application:

```
Could not load file or assembly 'System.Collections.Immutable, Version=1.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
```

Here are the nuget packages pulled in by the app:
```
PM> get-package | select id, version

Id                                                 Version     
--                                                 -------     
Antlr                                              3.4.1.9004  
bootstrap                                          3.0.0       
Google.Api.CommonProtos                            1.0.0       
Google.Api.Gax                                     1.0.1       
Google.Api.Gax.Grpc                                1.0.1       
Google.Apis                                        1.24.1      
Google.Apis.Auth                                   1.24.1      
Google.Apis.Core                                   1.24.1      
Google.Cloud.Diagnostics.AspNet                    1.0.0-beta09
Google.Cloud.Diagnostics.Common                    1.0.0-beta01
Google.Cloud.ErrorReporting.V1Beta1                1.0.0-beta05
Google.Cloud.Logging.Type                          1.0.0       
Google.Cloud.Logging.V2                            1.0.0       
Google.Cloud.Trace.V1                              1.0.0-beta05
Google.Protobuf                                    3.2.0       
Grpc.Auth                                          1.2.2       
Grpc.Core                                          1.2.2       
jQuery                                             1.10.2      
jQuery.Validation                                  1.11.1      
Microsoft.AspNet.Mvc                               5.2.3       
Microsoft.AspNet.Razor                             3.2.3       
Microsoft.AspNet.Web.Optimization                  1.1.3       
Microsoft.AspNet.WebApi.Client                     5.2.3       
Microsoft.AspNet.WebApi.Core                       5.2.3       
Microsoft.AspNet.WebPages                          3.2.3       
Microsoft.CodeDom.Providers.DotNetCompilerPlatform 1.0.0       
Microsoft.jQuery.Unobtrusive.Validation            3.2.3       
Microsoft.Net.Compilers                            1.0.0       
Microsoft.Web.Infrastructure                       1.0.0       
Modernizr                                          2.6.2       
Newtonsoft.Json                                    10.0.1      
Respond                                            1.2.0       
System.Collections.Immutable                       1.3.1       
System.Interactive.Async                           3.1.1       
System.Net.Http                                    4.0.0       
WebGrease                                          1.5.2       
Zlib.Portable.Signed                               1.11.0      
```

Here's the code I added:
```
    public class MvcApplication : System.Web.HttpApplication
    {
        //
        // Summary:
        //     Executes custom initialization code after all event handler modules have been
        //     added.
        public override void Init()
        {
            base.Init();
            CloudTrace.Initialize(this, "surferjeff-phoenix");
        }
```

The source code can be found here:
https://github.com/SurferJeffAtGoogle/csharp-docs-samples/tree/40f6be97cc70a281a4f1ff69a9fd95fc6dc74148/trace/api

### PR DESCRIPTION
**Pull request by evildour**
Fixed #1040

### CODE REVIEW

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs [hunk 1]
```diff
@@ -292,5 +291,37 @@ private ImmutableStack<TraceSpan> TraceStack
             }
         }
 #endif
+
+        // TODO: Remove when https://github.com/dotnet/roslyn/issues/12255 is fixed and go back to using System.Collections.Immutable
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Not sure even then, to be honest. If we can avoid an extra dependency just by having 30 lines of code, that could easily simplify things in the future.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs [hunk 2]
```diff
@@ -292,5 +291,37 @@ private ImmutableStack<TraceSpan> TraceStack
...
+        private class ImmutableStack<T>
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Sealed?

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs [hunk 3]
```diff
@@ -292,5 +291,37 @@ private ImmutableStack<TraceSpan> TraceStack
...
+        {
+            public static readonly ImmutableStack<T> Empty = new ImmutableStack<T>(default(T), null);
+
+            private ImmutableStack<T> _previous;
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Readonly for both variables?

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs (hunk 1 above)
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Good point. Removed.

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs (hunk 2 above)
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Done

File: apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs (hunk 3 above)
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs]: Well what if we want to mutate the immutable stack at some point? Just kidding. Done.

**Comment by evildour** [comment]
Thanks. Will merge when green.

**Comment by evildour** [comment]
@jskeet it looks like there's some error diffing files. Is this a known issue or something my change caused?

**Comment by jskeet** [comment]
Ah yes, bug in your change - dependencies are in `apis/apis.json`. CI noticed that regenerating the project files would have put the dependency back in. You need to remove it from `apis/apis.json`. (Another option is to remove it from there to start with and run `generateprojects.sh`.) Sorry that it's a bit convoluted, but it saves a lot of time in other ways :)

**Comment by evildour** [comment]
No that's ok, I just haven't been following too closely recently so I wasn't aware of that file. Let's see if this get's things working.

**Pull request merged**